<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Han Sun, Yue Wang, Zhen Xiao, Xiaoyun Huang, Haodong Wang, Tingting He, Xingpeng Jiang" />

<meta name="date" content="2022-11-22" />

<title>multiMiAT: An optimal microbiome-based association test for multicategory phenotypes</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">multiMiAT: An optimal microbiome-based association test for multicategory phenotypes</h1>
<h4 class="author">Han Sun, Yue Wang, Zhen Xiao, Xiaoyun Huang, Haodong Wang, Tingting He, Xingpeng Jiang</h4>
<h4 class="date">2022-11-22</h4>


<div id="TOC">
<ul>
<li><a href="#introduction">1. Introduction</a></li>
<li><a href="#installation">2. Installation</a><ul>
<li><a href="#requisite-r-packages">2.1. Requisite R packages</a></li>
<li><a href="#r-package-multimiat">2.2. R package <strong>multiMiAT</strong></a></li>
</ul></li>
<li><a href="#model">3. Model</a><ul>
<li><a href="#multinomial-logit-models">3.1. Multinomial logit models</a></li>
<li><a href="#microbiome-regression-based-kernel-association-test-for-multicategory-phenotypes-multimirkat">3.2. Microbiome regression-based kernel association test for multicategory phenotypes (multiMiRKAT)</a><ul>
<li><a href="#microbiome-regression-based-kernel-individual-tests">3.2.1. Microbiome regression-based kernel individual tests</a></li>
<li><a href="#microbiome-regression-based-kernel-omnibus-tests">3.2.2. Microbiome regression-based kernel omnibus tests</a></li>
</ul></li>
<li><a href="#score-test">3.3. Score test</a></li>
<li><a href="#mirkat-mc">3.4. MiRKAT-MC</a></li>
<li><a href="#optimal-test-for-multinomial-outcomes">3.5. Optimal test for multinomial outcomes</a></li>
</ul></li>
<li><a href="#implementation">4. Implementation</a></li>
<li><a href="#other-functions">5. Other functions</a><ul>
<li><a href="#function-simulateotu">5.1. Function <strong>SimulateOTU</strong></a><ul>
<li><a href="#description">5.1.1. Description</a></li>
<li><a href="#example">5.1.2. Example</a></li>
</ul></li>
<li><a href="#function-ordinal.simulated">5.2. Function <strong>ordinal.simulated</strong></a><ul>
<li><a href="#description-1">5.2.1. Description</a></li>
<li><a href="#example-1">5.2.2. Example</a></li>
</ul></li>
<li><a href="#function-nominal.simulated">5.3. Function <strong>nominal.simulated</strong></a><ul>
<li><a href="#description-2">5.3.1. Description</a></li>
<li><a href="#example-2">5.3.2. Example</a></li>
</ul></li>
<li><a href="#function-sigma_value">5.4. Function <strong>sigma_value</strong></a><ul>
<li><a href="#description-3">5.4.1. Description</a></li>
<li><a href="#example-3">5.4.2. Example</a></li>
</ul></li>
<li><a href="#function-kernel.matrix">5.5 Function <strong>Kernel.Matrix</strong></a><ul>
<li><a href="#description-4">5.5.1 Description</a></li>
<li><a href="#example-4">5.5.2 Example</a></li>
</ul></li>
<li><a href="#function-whole.kernels">5.6. Function <strong>Whole.Kernels</strong></a><ul>
<li><a href="#description-5">5.6.1. Description</a></li>
<li><a href="#example-5">5.6.2. Example</a></li>
</ul></li>
<li><a href="#function-multimirkat_bcl">5.7. Function <strong>multiMiRKAT_bcl</strong></a><ul>
<li><a href="#description-6">5.7.1 Description</a></li>
<li><a href="#example-6">5.7.2 Example</a></li>
</ul></li>
<li><a href="#function-multimirkat_clm">5.8. Function <strong>multiMiRKAT_clm</strong></a><ul>
<li><a href="#description-7">5.8.1 Description</a></li>
<li><a href="#example-7">5.8.2 Example</a></li>
</ul></li>
<li><a href="#function-multimirkat_crm">5.9. Function <strong>multiMiRKAT_crm</strong></a><ul>
<li><a href="#description-8">5.9.1. Description</a></li>
<li><a href="#example-8">5.9.2. Example</a></li>
</ul></li>
<li><a href="#function-score.test_bcl">5.10. Function <strong>score.test_bcl</strong></a><ul>
<li><a href="#description-9">5.10.1. Description</a></li>
<li><a href="#example-9">5.10.2. Example</a></li>
</ul></li>
<li><a href="#function-mirkat_mcn">5.11. Function <strong>MiRKAT_MCN</strong></a><ul>
<li><a href="#description-10">5.11.1. Description</a></li>
<li><a href="#example-10">5.11.2. Example</a></li>
</ul></li>
<li><a href="#function-mirkat_mco">5.12. Function <strong>MiRKAT_MCO</strong></a><ul>
<li><a href="#description-11">5.12.1. Description</a></li>
<li><a href="#example-11">5.12.2. Example</a></li>
</ul></li>
</ul></li>
<li><a href="#dataset">6. Dataset</a></li>
<li><a href="#references">7. References</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>1. Introduction</h1>
<p>This R package, <strong>multiMiAT</strong>, can be used for testing the association between microbiome and multicategory phenotypes. Specifically, <strong>multiMiAT</strong> can be applied to not only ordinal multicategory phenotypes (i.e., disease severity), but also nominal multicategory phenotypes (i.e., tumor subtype or dietary pattern). Considering that the binary outcome is a special case of the multicategory outcomes, our method can also be used for binary outcomes.</p>
</div>
<div id="installation" class="section level1">
<h1>2. Installation</h1>
<div id="requisite-r-packages" class="section level2">
<h2>2.1. Requisite R packages</h2>
<p>Our R package multiMiAT requires these existing R packages (“phyloseq”, “ecodist”, “GUniFrac”, “permute”, “ape”, “VGAM”, “MASS”, “Matrix”, “dirmult”, “harmonicmeanp”, “devtools”, “MiHC”) to be pre-imported, where “phyloseq” is available in Bioconductor, “ecodist”, “GUniFrac”, “permute”, “ape”, “VGAM”, “MASS”, “Matrix”, “dirmult”, “harmonicmeanp”, “devtools” are available in CRAN, and “MiHC” is available in github. We need to install these package using the following code:</p>
<pre><code>BiocManager::install(&quot;phyloseq&quot;)
install.packages(&quot;ecodist&quot;)
install.packages(&quot;GUniFrac&quot;)
install.packages(&quot;permute&quot;)
install.packages(&quot;ape&quot;)
install.packages(&quot;VGAM&quot;)
install.packages(&quot;MASS&quot;)
install.packages(&quot;Matrix&quot;)
install.packages(&quot;dirmult&quot;)
install.packages(&quot;harmonicmeanp&quot;)
install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;hk1785/MiHC&quot;)</code></pre>
</div>
<div id="r-package-multimiat" class="section level2">
<h2>2.2. R package <strong>multiMiAT</strong></h2>
<p>You may install <code>multiMiAT</code> from GitHub using the following code:</p>
<pre><code>devtools::install_github(&quot;xpjiang-ccnu/multiMiAT&quot;, force=T)</code></pre>
</div>
</div>
<div id="model" class="section level1">
<h1>3. Model</h1>
<div id="multinomial-logit-models" class="section level2">
<h2>3.1. Multinomial logit models</h2>
<p>For the observation value of the <span class="math inline">\(i\)</span>th subject, we record the multinomial response variable <span class="math inline">\(y_{i} \in \{1, 2, \ldots, J\}\)</span>, where <span class="math inline">\(i = 1, 2, \ldots, n\)</span> and <span class="math inline">\(J\)</span> &gt; 2. We suppose <span class="math inline">\(Y_{ij} = \text{I}\ (y_{i} = j)\)</span>, where <span class="math inline">\(\text{I}\ (\cdot)\)</span> is indicator function and <span class="math inline">\(j = 1, \ldots, J-1\)</span>, and define <span class="math inline">\(\textbf{Y}_{i} = (Y_{i1}, \ldots, Y_{i(J-1)})^T\)</span> as response vector. We also observe <span class="math inline">\(m\)</span> OTUs abundance <span class="math inline">\(\textbf{o}_{i} = (o_{i1}, o_{i2}, \ldots, o_{im})^T\)</span> and <span class="math inline">\(t\)</span> covariate <span class="math inline">\(\textbf{x}_{i} = (x_{i1}, \ldots, x_{it})^T\)</span>. We assume that the marginal expectation of <span class="math inline">\(Y_{ij}\)</span> given <span class="math inline">\(\textbf{x}_{i},\textbf{o}_{i}\)</span> is <span class="math inline">\(\mu_{ij}:= E(Y_{ij}|\textbf{x}_{i},\textbf{o}_{i}) = P(y_{i} = j|\textbf{x}_{i},\textbf{o}_{i})\)</span>, we connect the marginal expected vector and covariates by specifying that</p>
<p><span class="math display">\[\begin{equation}
    g(\mu_{ij}) = \alpha_{0j} + \textbf{x}_{i}^T {\alpha} + f({\textbf{o}}_{i}),\label{logit}
\end{equation}\]</span> where <span class="math inline">\(\alpha_{0j}\)</span> and <span class="math inline">\(\alpha\)</span> are the intercept and regression coefficients, respectively. <span class="math inline">\(g(\cdot)\)</span> is a link function. <span class="math inline">\(f(\cdot)\)</span> is a function reflecting the microbiome profile. We consider the cumulative link model (clm, also called the proportional odds model) and baseline category logit model (bcl) as the link function, respectively, whose forms are as follows: <span class="math display">\[\begin{equation}
    \text{log}(\frac{P(y_{i} \leq j| \textbf{x}_{i}, \textbf{o}_{i})}{1-P(y_{i} \leq j| \textbf{x}_{i}, \textbf{o}_{i})}) = \alpha_{0j} + \textbf{x}_{i}^T \alpha_{\ast} + f({\textbf{o}}_{i}),\label{clm1}
\end{equation}\]</span> <span class="math display">\[\begin{equation}
    \text{log}(\frac{\mu_{ij}}{\mu_{iJ}}) = \alpha_{0j} + \textbf{x}_{i}^T \alpha_{j} + f({\textbf{o}}_{i}).\label{bcl1}
\end{equation}\]</span> Here, the regression coefficients <span class="math inline">\(\alpha_{\ast}\)</span> are fixed for all <span class="math inline">\(j\)</span>. <span class="math inline">\(\alpha_{0j}\)</span> is the category-specific intercept and needs to satisfy a monotonicity constraint (i.e., <span class="math inline">\(\alpha_{01} \leq \ldots \leq \alpha_{0(J-1)}\)</span>) only in the cumulative link model. <span class="math inline">\(\alpha_{j}\)</span> is the category-specific regression coefficient.</p>
</div>
<div id="microbiome-regression-based-kernel-association-test-for-multicategory-phenotypes-multimirkat" class="section level2">
<h2>3.2. Microbiome regression-based kernel association test for multicategory phenotypes (multiMiRKAT)</h2>
<div id="microbiome-regression-based-kernel-individual-tests" class="section level3">
<h3>3.2.1. Microbiome regression-based kernel individual tests</h3>
<p>We adopt seven types of distances to calculate the dissimilarity matrix between subjects. Specifically, Bray-Curtis dissimilarity (<span class="math inline">\(\text{D}_{\text{BC}}\)</span>), as a non-phylogeny-based dissimilarity, and the unweighted and weighted UniFrac (<span class="math inline">\(\text{D}_{\text{u}}\)</span> and <span class="math inline">\(\text{D}_{\text{w}}\)</span>), as the phylogeny-based distance, are taken into account, which are also commonly used in microbial research. Considering weight assignments, we adopt the four generalized UniFrac with <span class="math inline">\(\alpha\)</span> = 0, 0.25, 0.5, and 0.75 (<span class="math inline">\(\text{D}_{\text{0}}\)</span>, <span class="math inline">\(\text{D}_{\text{0.25}}\)</span>, <span class="math inline">\(\text{D}_{\text{0.5}}\)</span> and <span class="math inline">\(\text{D}_{\text{0.75}}\)</span>). We convert the distance matrix into a kernel matrix via a kernel function. Here, we consider taking linear and nonlinear kernel functions into our model, whose models are as follows:</p>
<p><span class="math inline">\(\textbf{Linear kernel} (\text{K}_{\text{LK}}): \textbf{K}_{d} = - \frac{1}{2} (\textbf{I}_n - \frac {\textbf{1}_n \textbf{1}^T_n}{n}) \textbf{D}_d^2 (\textbf{I}_n - \frac {\textbf{1}_n \textbf{1}^T_n}{n}),\)</span></p>
<p><span class="math inline">\(\textbf{Gaussian kernel} (\text{K}_{\text{GK}}): \textbf{K} = \text{exp} (- \frac{\textbf{D}_{d}^2}{2 \sigma^2}),\)</span></p>
<p><span class="math inline">\(\textbf{Laplacian kernel} (\text{K}_{\text{LaK}}): \textbf{K} = \text{exp} (- \frac{\textbf{D}_{d}}{\sigma}),\)</span></p>
<p>where <span class="math inline">\(\textbf{I}_n\)</span> is a <span class="math inline">\(n \times n\)</span> identity matrix. <span class="math inline">\(\textbf{1}_n\)</span> is a <span class="math inline">\(n\)</span>-dimensional vector of ones. <span class="math inline">\(\textbf{D}_{d}\)</span> is dissimilarity matrix and <span class="math inline">\(d \in \Gamma_{\text{D}} = \{\text{D}_{\text{BC}}, \text{D}_{\text{u}}, \text{D}_{\text{0}}, \text{D}_{\text{0.25}},\)</span> <span class="math inline">\(\text{D}_{\text{0.5}}, \text{D}_{\text{0.75}}, \text{D}_{\text{w}}\}\)</span>. <span class="math inline">\(\sigma\)</span> is the hyperparameter. We also conduct the positive semi-definiteness correction procedure for <span class="math inline">\(\textbf{K}\)</span> to ensure that the eigenvalue of <span class="math inline">\(\textbf{K}\)</span> is non-negative. We calculate the eigenvalues <span class="math inline">\(\lambda_i\)</span> (<span class="math inline">\(i = 1, \ldots, n\)</span>) and eigenvectors of <span class="math inline">\(\textbf{K}\)</span> via eigenvalue decomposition <span class="math inline">\(\textbf{K} = \textbf{U} {\Lambda} \textbf{U}^T\)</span>, where <span class="math inline">\({\Lambda}\)</span> is the diagonal matrix with the <span class="math inline">\(i\)</span>th diagonal element <span class="math inline">\(\lambda_i\)</span>. Then, we obtain the reconstructed kernel matrix <span class="math inline">\(\textbf{K}^* = \textbf{U} \Lambda^* \textbf{U}^T\)</span>, where <span class="math inline">\(\Lambda^*\)</span> is the diagonal matrix with the <span class="math inline">\(i\)</span>th diagonal element <span class="math inline">\(\left | \lambda_i \right |\)</span>.</p>
<p>We define <span class="math inline">\(\textbf{Y} = (\textbf{Y}_{1}^T, \ldots \textbf{Y}_{n}^T)^T\)</span> and the test statistic of individual tests are <span class="math display">\[\begin{equation}
    T_l^{kd} = (\textbf{Y} - \hat{{\mu}}_l)^T \textbf{V}^{-1}_{l} {\widetilde{\textbf{K}}}_{kd} \textbf{V}^{-1}_{l} (\textbf{Y} - \hat{{\mu}}_l),
\end{equation}\]</span> where <span class="math inline">\(\hat{{\mu}}_l = (\hat{\mu}_l^{11}, \ldots, \hat{\mu}_l^{1(J-1)}, \ldots, \hat{\mu}_l^{n1}, \ldots, \mu_l^{n(J-1)})^T\)</span>. <span class="math inline">\(\hat{{\mu}}_l\)</span> is the the estimated expectation via multinomial logit model under <span class="math inline">\(H_0\)</span> condition, where <span class="math inline">\(l \in \Gamma_{\text{L}} = \{\text{M}_\text{bcl}, \text{M}_\text{clm}\}\)</span>. <span class="math inline">\(\textbf{V}_{l} = \textbf{I}_n \otimes \textbf{V}_{l0}\)</span> and <span class="math inline">\(\textbf{V}_{l0}\)</span> is the estimated residual variance matrix under the null model. <span class="math inline">\({\widetilde{\textbf{K}}_{kd}} = \textbf{K}^*_{kd} \otimes \textbf{1}_{(J-1) \times (J-1)}\)</span>, where <span class="math inline">\(\textbf{1}_{(J-1) \times (J-1)}\)</span> is <span class="math inline">\((J-1) \times (J-1)\)</span> matrix of ones, <span class="math inline">\(k \in \Gamma_{\text{K}} = \{\text{K}_{\text{LK}}, \text{K}_{\text{GK}}, \text{K}_{\text{LaK}}\}\)</span>.</p>
</div>
<div id="microbiome-regression-based-kernel-omnibus-tests" class="section level3">
<h3>3.2.2. Microbiome regression-based kernel omnibus tests</h3>
<p>We take the minimum <span class="math inline">\(p\)</span>-values of individual tests as the test statistics of global omnibus tests <span class="math display">\[\begin{equation}
    T_{\text{multiMiRKAT}} = \underset{k \in \Gamma_{\text{K}}}{\text{min}} \underset{d \in \Gamma_{\text{D}}}{\text{min}} p_l^{kd}.
\end{equation}\]</span> where <span class="math inline">\(p_l^{kd}\)</span> denotes the <span class="math inline">\(p\)</span>-value of <span class="math inline">\(T_l^{kd}\)</span>. Considering that the regression parameter coefficients fitted ordinal/nominal multinomial logit model may be inconsistent, we adopt two multinomial logit models (i.e., clm and bcl) to model the global omnibus tests <span class="math inline">\(T_{\text{multiMiRKAT}}\)</span>. Specifically, we recommend using clm/bcl to construct the test statistic of global omnibus test for ordinal/nominal multinomial outcomes, respectively. We call the global omnibus test for ordinal or nominal multinomial outcomes multiMiRKAT-O or multiMiRKAT-N.</p>
</div>
</div>
<div id="score-test" class="section level2">
<h2>3.3. Score test</h2>
<p>We define that <span class="math inline">\(\textbf{O} = (\textbf{o}_{1}, \ldots, \textbf{o}_{n})^T\)</span> and <span class="math inline">\(\textbf{W}\)</span> is a <span class="math inline">\(m\)</span>-dimensional weight vector which represents the weights of <span class="math inline">\(m\)</span> OTUs. Let <span class="math inline">\(\textbf{S} = \textbf{I}_{(J-1)} \otimes (\textbf{OW}^T)\)</span>, where <span class="math inline">\(\textbf{I}_{(J-1)}\)</span> represents a <span class="math inline">\((J-1) \times (J-1)\)</span> identity matrix and <span class="math inline">\(\otimes\)</span> is the Kronecker product. Here, we adopt <span class="math inline">\(\textbf{W} = \textbf{1}_m\)</span> and <span class="math inline">\(\textbf{1}_m\)</span> is a <span class="math inline">\(m\)</span>-dimensional vector of ones. Let <span class="math inline">\(\overline{\textbf{Y}}_{j} = (Y_{1j}, \ldots, Y_{nj})^T\)</span> and <span class="math inline">\(\overline{\textbf{Y}} = (\overline{\textbf{Y}}_{1}^T, \ldots, \overline{\textbf{Y}}_{m}^T)^T\)</span>. The covariance matrix of <span class="math inline">\(\overline{\textbf{Y}}\)</span> is <span class="math display">\[\begin{equation}
{\Sigma} = \begin{pmatrix}

    {\Sigma}_{1,1} &amp; \cdots &amp; {\Sigma}_{1, J - 1} \\
    
    \vdots &amp; \ddots &amp; \vdots \\
    
    {\Sigma}_{J - 1, 1} &amp; \cdots &amp; {\Sigma}_{J - 1, J - 1}
    
\end{pmatrix},
\end{equation}\]</span> where <span class="math display">\[\begin{equation}
  {\Sigma}_{j,j&#39;} = \left\{
    \begin{aligned}
        \text{diag}({\mu}_{1j}(1-{\mu}_{1j}), \ldots, {\mu}_{nj}(1-{\mu}_{nj}))  &amp;\qquad \text{if} \quad j = j&#39; \\
        \text{diag}(- {\mu}_{1j}{\mu}_{1j&#39;}, \ldots, - {\mu}_{nj}{\mu}_{nj&#39;}) \quad &amp;\qquad \text{if} \quad j \neq j&#39;
    \end{aligned}
\right..
\end{equation}\]</span> We define <span class="math inline">\(\overline{{\mu}} = (\overline{{\mu}}_{1}^T, \ldots, \overline{{\mu}}_{m}^T)^T\)</span> and <span class="math inline">\(\textbf{U} = \textbf{S}^T (\overline{\textbf{Y}} - \overline{{\mu}})\)</span>, where <span class="math inline">\(\overline{{\mu}}_{j} = (\mu_{1j}, \ldots, \mu_{nj})^T\)</span>. Let <span class="math inline">\(\textbf{X} = (\textbf{x}_{1}, \ldots, \textbf{x}_{n})^T\)</span> and <span class="math inline">\(\mathbb{X} = \textbf{I}_{(J-1)} \otimes \textbf{X}\)</span>. Then we establish test statistic of score test <span class="math display">\[\begin{equation}
    T_{\text{score}} = \textbf{U}^T \widehat{{\Sigma}}^{-1} \textbf{U}
\end{equation}\]</span> where <span class="math inline">\(\widehat{{\Sigma}} = \textbf{S}^T ({\Sigma} - {\Sigma} \mathbb{X} (\mathbb{X}^T {\Sigma} \mathbb{X})^{-1} \mathbb{X}^T {\Sigma}) \textbf{S})\)</span> is the covariance matrix of <span class="math inline">\(\textbf{U}\)</span>. <span class="math inline">\(T_{\text{score}}\)</span> can be shown to asymptotically follow a <span class="math inline">\(\chi^2\)</span> distribution with degree of freedom <span class="math inline">\((J - 1)\)</span>.</p>
</div>
<div id="mirkat-mc" class="section level2">
<h2>3.4. MiRKAT-MC</h2>
<p>MiRKAT-MC is proposed for multicategory phenotypes. Similar to the OMiRKAT and all its extensions, MiRKAT-MC also utilizes the <span class="math inline">\(p\)</span>-value combination method (i.e., HMP) to integrate individual tests. Here, the test statistic of individual tests is</p>
<p><span class="math display">\[\begin{equation}
    Q_l^{kd} = (\overline{\textbf{Y}} - \overline{\mu}_l)^T \textbf{H}_l^T \overline{\textbf{W}}_l \thinspace {\overline{\textbf{K}}}_{kd} \overline{\textbf{W}}_l \thinspace \textbf{H}_l (\overline{\textbf{Y}} - \overline{\mu}_l),
\end{equation}\]</span> where <span class="math inline">\({\overline{\textbf{K}}}_{kd} = \textbf{I}_{(J-1)} \otimes \textbf{K}^*_{kd}\)</span> and <span class="math inline">\(\overline{\textbf{W}}_l = (\textbf{H}_l \overline{\textbf{V}}_l \textbf{H}_l)^{-1}\)</span>. <span class="math inline">\(\textbf{H}_l = \partial (\mathbb{X} \hat{\beta})/\partial \overline{\mu}_l\)</span> and <span class="math inline">\(\overline{\textbf{V}}_l\)</span> is the variance-covariance matrix. MiRKAT-MC test consists of bcl-based MiRKAT-MCN test and clm-based MiRKAT-MCO test.</p>
</div>
<div id="optimal-test-for-multinomial-outcomes" class="section level2">
<h2>3.5. Optimal test for multinomial outcomes</h2>
<p>The optimal test multiMiAT adopts HMP to cpmbine the <span class="math inline">\(p\)</span>-values of multiMiRKAT test, score test and MiRKAT-MC test, whose test statistic is <span class="math display">\[\begin{equation}
    T_{\text{multiMiAT}} = \frac{\sum_{t \in \Gamma_\text{T}} w_t} {\sum_{t \in \Gamma_\text{T}} w_t / p_t},
\end{equation}\]</span> where <span class="math inline">\(w_t\)</span> . <span class="math inline">\(\Gamma_\text{T} = \{T_{\text{multiMiRKAT-N}}, T_{\text{multiMiRKAT-O}}, T_{\text{score}}, T_{\text{MiRKAT-MCN}}, T_{\text{MiRKAT-MCO}}\}\)</span> and <span class="math inline">\(p_t\)</span> represents the <span class="math inline">\(p\)</span>-values of these combined tests. Remarkably, the MiRKAT-MC test combined in our method is with more distances and kernel functions compared to the original MiRKAT-MC test. Considering the advantages of different association tests, the optimal test multiMiAT can accommodate diverse association patterns.</p>
</div>
</div>
<div id="implementation" class="section level1">
<h1>4. Implementation</h1>
<p><strong>First of all, we import requisite R packages:</strong></p>
<pre><code>require(ecodist)
require(GUniFrac)
require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Then, we import example microbiome data:</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])
W &lt;- matrix(1, nrow = 1, ncol = ncol(otu.tab))</code></pre>
<p><strong>The microbiome data we use includes OTU count table, phylogenetic tree and sample information, where sample information contains response variable (e.g., disease status) and covariates (e.g., age, gender). Next, we test the association between CRC status and microbial composition using multiMiAT:</strong></p>
<pre><code>out &lt;- multiMiAT(y = y, otu.tab = otu.tab, tree = tree, covs = covs, method = c(&quot;multiMiRKAT&quot;, &quot;score&quot;, &quot;MiRKAT-MC&quot;), distance = c(&quot;BC&quot;, &quot;U.UniFrac&quot;, &quot;G.UniFrac.00&quot;, &quot;G.UniFrac.25&quot;, &quot;G.UniFrac.50&quot;, &quot;G.UniFrac.75&quot;, &quot;W.UniFrac&quot;), kernel = c(&quot;LK&quot;, &quot;GK&quot;, &quot;LaK&quot;), W = W)
out</code></pre>
<pre><code>## $multiMiRKAT.pvs
## $multiMiRKAT.pvs$multiMiRKAT.bcl.pvs
## $multiMiRKAT.pvs$multiMiRKAT.bcl.pvs$LK
##           BC    U.UniFrac G.UniFrac.00 G.UniFrac.25 G.UniFrac.50 G.UniFrac.75 
##       0.0761       0.0060       0.0581       0.1211       0.2352       0.3003 
##    W.UniFrac  Omni.MiRKAT 
##       0.2643       0.0230 
## 
## $multiMiRKAT.pvs$multiMiRKAT.bcl.pvs$GK.sigma1
##           BC    U.UniFrac G.UniFrac.00 G.UniFrac.25 G.UniFrac.50 G.UniFrac.75 
##       0.0871       0.0070       0.0601       0.1291       0.2543       0.3283 
##    W.UniFrac  Omni.MiRKAT 
##       0.2993       0.0270 
## 
## $multiMiRKAT.pvs$multiMiRKAT.bcl.pvs$LaK.sigma1
##           BC    U.UniFrac G.UniFrac.00 G.UniFrac.25 G.UniFrac.50 G.UniFrac.75 
##       0.1221       0.0080       0.0721       0.1582       0.3844       0.5385 
##    W.UniFrac  Omni.MiRKAT 
##       0.5906       0.0270 
## 
## 
## $multiMiRKAT.pvs$multiMiRKAT.clm.pvs
## $multiMiRKAT.pvs$multiMiRKAT.clm.pvs$LK
##           BC    U.UniFrac G.UniFrac.00 G.UniFrac.25 G.UniFrac.50 G.UniFrac.75 
##       0.0651       0.0010       0.0380       0.0961       0.2142       0.3073 
##    W.UniFrac  Omni.MiRKAT 
##       0.3143       0.0050 
## 
## $multiMiRKAT.pvs$multiMiRKAT.clm.pvs$GK.sigma1
##           BC    U.UniFrac G.UniFrac.00 G.UniFrac.25 G.UniFrac.50 G.UniFrac.75 
##       0.0851       0.0050       0.0511       0.1131       0.2663       0.4364 
##    W.UniFrac  Omni.MiRKAT 
##       0.5145       0.0170 
## 
## $multiMiRKAT.pvs$multiMiRKAT.clm.pvs$LaK.sigma1
##           BC    U.UniFrac G.UniFrac.00 G.UniFrac.25 G.UniFrac.50 G.UniFrac.75 
##       0.1281       0.0070       0.0521       0.1351       0.3684       0.5746 
##    W.UniFrac  Omni.MiRKAT 
##       0.7067       0.0240 
## 
## 
## 
## $Combined.methods.pvs
## multiMiRKAT-N multiMiRKAT-O    score test    MiRKAT-MCN    MiRKAT-MCO 
##        0.0290        0.0060        0.0835        0.0401        0.0232 
## 
## $Optimal.test.p
## multiMiAT 
##    0.0197</code></pre>
<p><strong>We can find significant association between CRC status and microbial composition based on multiMiAT (<em>p</em>-value = 0.0197 &lt; 0.05).</strong></p>
</div>
<div id="other-functions" class="section level1">
<h1>5. Other functions</h1>
<div id="function-simulateotu" class="section level2">
<h2>5.1. Function <strong>SimulateOTU</strong></h2>
<div id="description" class="section level3">
<h3>5.1.1. Description</h3>
<p>The function, <strong>SimulateOTU</strong>, generates the OTU count table simulated based on the Dirichlet-multinomial model according to real data, which is the same as the <strong>SimulateOTU</strong> function of the package <strong>GEEMiHC</strong>. We first calculate the parameters of the real data, and then generate the simulated data according to the parameters.</p>
</div>
<div id="example" class="section level3">
<h3>5.1.2. Example</h3>
<p><strong>Parameter estimation:</strong></p>
<pre><code>require(phyloseq)
require(multiMiAT)
data(&quot;phy&quot;, package = &quot;MiHC&quot;)
phy.otu.tab &lt;- otu_table(phy)
# require(dirmult)
# phy.parameters &lt;- dirmult(phy.otu.tab)
data(&quot;phy.parameters&quot;)</code></pre>
<p><strong>Generation of microbiome data:</strong></p>
<pre><code>otu.tab &lt;- SimulateOTU(phy.otu.tab, nSam = 100, phy.parameters, mu = 1000, size = 25)</code></pre>
</div>
</div>
<div id="function-ordinal.simulated" class="section level2">
<h2>5.2. Function <strong>ordinal.simulated</strong></h2>
<div id="description-1" class="section level3">
<h3>5.2.1. Description</h3>
<p>The function, <strong>ordinal.simulated</strong>, generates the simulated data including OTU count table, ordinal multinomial host phenotype (e.g., disease severity) and confounding factors (e.g., age, gender), which mainly refers to the <strong>rmult.clm</strong> function in R package, <strong>SimCorMultRes</strong>.</p>
</div>
<div id="example-1" class="section level3">
<h3>5.2.2. Example</h3>
<p><strong>Parameter settings:</strong></p>
<pre><code>sample_size &lt;- 100
OTU_select &lt;- 30
betas0 &lt;- c(0.5, 0.5)
beta &lt;- 1</code></pre>
<p><strong>Generation of microbiome data:</strong></p>
<pre><code>require(phyloseq)
require(multiMiAT)
data(&quot;phy&quot;, package = &quot;MiHC&quot;)
phy.otu.tab &lt;- otu_table(phy)
tree &lt;- phy_tree(phy)
# library(dirmult)
# phy.parameters &lt;- dirmult(phy.otu.tab)
data(&quot;phy.parameters&quot;)
otu.tab &lt;- SimulateOTU(phy.otu.tab, nSam = 100, phy.parameters, mu = 1000, size = 25)</code></pre>
<p><strong>Random selection of OTUs:</strong></p>
<pre><code>otu.tab_selection &lt;- otu.tab[, order(apply(otu.tab, 2, sum), decreasing = T)[1:OTU_select]]
colnames(otu.tab_selection) &lt;- paste(&quot;OTU&quot;, colnames(otu.tab_selection), sep = &quot;&quot;)
otu.tab_selection &lt;- scale(otu.tab_selection)</code></pre>
<p><strong>Generation of covariates and regression parameters:</strong></p>
<pre><code>betas1 &lt;- rep(beta, ncol(otu.tab_selection))
x1 &lt;- rbinom(sample_size, 1, 0.5)
x2 &lt;- rnorm(sample_size, 0, 1)
x1 &lt;- scale(x1)
x2 &lt;- scale(x2)
xdata &lt;- data.frame(x1, x2, otu.tab_selection)
beta_coefficients &lt;- c(betas0, betas1)
xnam &lt;- colnames(xdata)
xformula &lt;- as.formula(paste(&quot;~ &quot;, paste(xnam, collapse= &quot;+&quot;)))
covs &lt;- data.frame(x1, x2)
covs[,1] &lt;- as.factor(covs$x1) 
covs[,2] &lt;- as.numeric(covs$x2)</code></pre>
<p><strong>Generation of ordinal multinomial outcomes:</strong></p>
<pre><code>y &lt;- ordinal.simulated(betas = beta_coefficients, proportion = c(1, 1, 1, 1), xformula = xformula, xdata = xdata, link = &quot;logit&quot;) </code></pre>
<p><strong>Data synthesis of simulated data with ordinal multinomial outcome:</strong></p>
<pre><code>otu.tab_information &lt;- otu_table(otu.tab, taxa_are_rows = F)
sample_information &lt;- sample_data(as.data.frame(cbind(y, x1, x2)))
colnames(sample_information) &lt;- c(&quot;y&quot;, &quot;x1&quot;, &quot;x2&quot;)
rownames(sample_information) &lt;- rownames(otu.tab_information)
biom &lt;- phyloseq(otu.tab_information, tree, sample_information)</code></pre>
</div>
</div>
<div id="function-nominal.simulated" class="section level2">
<h2>5.3. Function <strong>nominal.simulated</strong></h2>
<div id="description-2" class="section level3">
<h3>5.3.1. Description</h3>
<p>The function, <strong>nominal.simulated</strong>, generates the simulated data including OTU count table, nominal multinomial host phenotype (e.g., tumor subtype) and confounding factors (e.g., age, gender), which mainly refers to the <strong>rmult.bcl</strong> function in R package, <strong>SimCorMultRes</strong>.</p>
</div>
<div id="example-2" class="section level3">
<h3>5.3.2. Example</h3>
<p><strong>Parameter settings:</strong></p>
<pre><code>sample_size &lt;- 100
categories &lt;- 4
OTU_select &lt;- 30
betas0 &lt;- c(1, 0.5, 0.5)
beta &lt;- 1</code></pre>
<p><strong>Generation of microbiome data:</strong></p>
<pre><code>require(phyloseq)
require(multiMiAT)
data(&quot;phy&quot;, package = &quot;MiHC&quot;)
phy.otu.tab &lt;- otu_table(phy)
tree &lt;- phy_tree(phy)
# library(dirmult)
# phy.parameters &lt;- dirmult(phy.otu.tab)
data(&quot;phy.parameters&quot;)
otu.tab &lt;- SimulateOTU(phy.otu.tab, nSam = 100, phy.parameters, mu = 1000, size = 25)</code></pre>
<p><strong>Random selection of OTUs:</strong></p>
<pre><code>otu.tab_selection &lt;- otu.tab[, order(apply(otu.tab, 2, sum), decreasing = TRUE)[1:OTU_select]]
colnames(otu.tab_selection) &lt;- paste(&quot;OTU&quot;, colnames(otu.tab_selection), sep = &quot;&quot;)
otu.tab_selection &lt;- scale(otu.tab_selection)</code></pre>
<p><strong>Generation of covariates and regression parameters:</strong></p>
<pre><code>x1 &lt;- rbinom(sample_size, 1, 0.5)
x2 &lt;- rnorm(sample_size, 0, 1)
x1 &lt;- scale(x1)
x2 &lt;- scale(x2)
xdata &lt;- data.frame(x1, x2, otu.tab_selection)
beta_coefficients &lt;- c()
for (i in 1:(categories-1)) {
  betas1 &lt;- round(runif(ncol(otu.tab_selection), 0.1, beta), digits = 1)
  beta_coefficients &lt;- c(beta_coefficients, betas0, betas1)
}
beta_coefficients &lt;- c(beta_coefficients, rep(0, (ncol(otu.tab_selection) + 3)))
xnam &lt;- colnames(xdata)
xformula &lt;- as.formula(paste(&quot;~ &quot;, paste(xnam, collapse= &quot;+&quot;)))</code></pre>
<p><strong>Generation of nomial multinomial outcomes:</strong></p>
<pre><code>y &lt;- nominal.simulated(betas=beta_coefficients, categories=categories, xformula = xformula, xdata = xdata) </code></pre>
<p><strong>Data synthesis of simulated data with nomial multinomial outcome:</strong></p>
<pre><code>otu.tab_information &lt;- otu_table(otu.tab, taxa_are_rows = FALSE)
sample_information &lt;- sample_data(as.data.frame(cbind(y, x1, x2)))
colnames(sample_information) &lt;- c(&quot;y&quot;, &quot;x1&quot;, &quot;x2&quot;)
rownames(sample_information) &lt;- rownames(otu.tab_information)
biom &lt;- phyloseq(otu.tab_information, tree, sample_information)</code></pre>
</div>
</div>
<div id="function-sigma_value" class="section level2">
<h2>5.4. Function <strong>sigma_value</strong></h2>
<div id="description-3" class="section level3">
<h3>5.4.1. Description</h3>
<p>The median value of distance matrix square (D * D) or distance matrix (D).</p>
</div>
<div id="example-3" class="section level3">
<h3>5.4.2. Example</h3>
<pre><code>require(GUniFrac)
require(multiMiAT)
data(throat.otu.tab)
data(throat.tree)
unifs &lt;- GUniFrac::GUniFrac(throat.otu.tab, throat.tree, alpha = c(0, 0.25, 0.5, 0.75, 1))$unifracs
u.unif &lt;- unifs[, , &quot;d_UW&quot;]
sigma.u.unif.list &lt;- sigma_value(u.unif)</code></pre>
</div>
</div>
<div id="function-kernel.matrix" class="section level2">
<h2>5.5 Function <strong>Kernel.Matrix</strong></h2>
<div id="description-4" class="section level3">
<h3>5.5.1 Description</h3>
<p>The function, <strong>Kernel.Matrix</strong>, establishes the kernel matrix through the kernel function (i.e., linear kernel, Gaussian kernel, exponential kernel and Laplacian kernel function).</p>
</div>
<div id="example-4" class="section level3">
<h3>5.5.2 Example</h3>
<pre><code>require(GUniFrac)
require(multiMiAT)
data(throat.otu.tab)
data(throat.tree)
unifs &lt;- GUniFrac::GUniFrac(throat.otu.tab, throat.tree, alpha = c(0, 0.25, 0.5, 0.75, 1))$unifracs
u.unif &lt;- unifs[, , &quot;d_UW&quot;]
u.unif.k &lt;- Kernel.Matrix(u.unif, kernel = c(&quot;LK&quot;, &quot;GK&quot;, &quot;EK&quot;, &quot;LaK&quot;), sigma.GK = 1, sigma.EK = 1, sigma.LaK = 1)</code></pre>
</div>
</div>
<div id="function-whole.kernels" class="section level2">
<h2>5.6. Function <strong>Whole.Kernels</strong></h2>
<div id="description-5" class="section level3">
<h3>5.6.1. Description</h3>
<p>The construction of the kernel matrix through diverse distances (i.e., Jaccard dissimilarity, Bray-Curtis distance, Unifrac distance, Generalized UniFrac distance with <span class="math inline">\(\alpha\)</span> = 0, Generalized UniFrac distance with <span class="math inline">\(\alpha\)</span> = 0.25, Generalized UniFrac distance with <span class="math inline">\(\alpha\)</span> = 0.5, Generalized UniFrac distance with <span class="math inline">\(\alpha\)</span> = 0.75 and Weighted UniFrac distance) and kernel functions (i.e., linear kernel, Gaussian kernel, exponential kernel and Laplacian kernel function).</p>
</div>
<div id="example-5" class="section level3">
<h3>5.6.2. Example</h3>
<pre><code>require(GUniFrac)
require(ecodist)
require(multiMiAT)
data(throat.otu.tab)
data(throat.tree)
Ks &lt;- Whole.Kernels(throat.otu.tab, throat.tree) </code></pre>
</div>
</div>
<div id="function-multimirkat_bcl" class="section level2">
<h2>5.7. Function <strong>multiMiRKAT_bcl</strong></h2>
<div id="description-6" class="section level3">
<h3>5.7.1 Description</h3>
<p>The function, <strong>multiMiRKAT_bcl</strong>, can be used to detect the association between the microbiome and multinomial outcomes, which is microbiome regression-based kernel association tests based on baseline category logit model.</p>
</div>
<div id="example-6" class="section level3">
<h3>5.7.2 Example</h3>
<p><strong>Import requisite R packages</strong></p>
<pre><code>require(ecodist)
require(GUniFrac)
require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Import example microbiome data</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])</code></pre>
<p><strong>Fit multiMiRKAT_bcl</strong></p>
<pre><code>Ks &lt;- Whole.Kernels(otu.tab, tree) 
out &lt;- multiMiRKAT_bcl(y = y, Ks = Ks, covs = covs)</code></pre>
</div>
</div>
<div id="function-multimirkat_clm" class="section level2">
<h2>5.8. Function <strong>multiMiRKAT_clm</strong></h2>
<div id="description-7" class="section level3">
<h3>5.8.1 Description</h3>
<p>The function, <strong>multiMiRKAT_clm</strong>, can be used to detect the association between the microbiome and multinomial outcomes, which is microbiome regression-based kernel association tests based on cumulative link model.</p>
</div>
<div id="example-7" class="section level3">
<h3>5.8.2 Example</h3>
<p><strong>Import requisite R packages</strong></p>
<pre><code>require(ecodist)
require(GUniFrac)
require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Import example microbiome data</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])</code></pre>
<p><strong>Fit multiMiRKAT_clm</strong></p>
<pre><code>Ks &lt;- Whole.Kernels(otu.tab, tree) 
out &lt;- multiMiRKAT_clm(y = y, Ks = Ks, covs = covs)</code></pre>
</div>
</div>
<div id="function-multimirkat_crm" class="section level2">
<h2>5.9. Function <strong>multiMiRKAT_crm</strong></h2>
<div id="description-8" class="section level3">
<h3>5.9.1. Description</h3>
<p>The function, <strong>multiMiRKAT_crm</strong>, can be used to detect the association between the microbiome and multinomial outcomes, which is microbiome regression-based kernel association tests based on continuation ratio model.</p>
</div>
<div id="example-8" class="section level3">
<h3>5.9.2. Example</h3>
<p><strong>Import requisite R packages</strong></p>
<pre><code>require(ecodist)
require(GUniFrac)
require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Import example microbiome data</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])</code></pre>
<p><strong>Fit multiMiRKAT_crm</strong></p>
<pre><code>Ks &lt;- Whole.Kernels(otu.tab, tree) 
out &lt;- multiMiRKAT_crm(y = y, Ks = Ks, covs = covs)</code></pre>
</div>
</div>
<div id="function-score.test_bcl" class="section level2">
<h2>5.10. Function <strong>score.test_bcl</strong></h2>
<div id="description-9" class="section level3">
<h3>5.10.1. Description</h3>
<p>The function, <strong>score.test_bcl</strong>, is score test based on baseline category logit model, which is a special case of an individual test in subtype analysis with somatic mutations (SASOM). It can be also used for the association between the microbiome and multinomial outcomes.</p>
</div>
<div id="example-9" class="section level3">
<h3>5.10.2. Example</h3>
<p><strong>Import requisite R packages</strong></p>
<pre><code>require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Import example microbiome data</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])
W &lt;- matrix(1, nrow = 1, ncol = ncol(otu.tab))</code></pre>
<p><strong>Fit bcl-based score test</strong></p>
<pre><code>out &lt;- score.test_bcl(y, otu.tab, covs, W)</code></pre>
</div>
</div>
<div id="function-mirkat_mcn" class="section level2">
<h2>5.11. Function <strong>MiRKAT_MCN</strong></h2>
<div id="description-10" class="section level3">
<h3>5.11.1. Description</h3>
<p>The function, <strong>MiRKAT_MCN</strong>, is MiRKAT_MC based on baseline category logit model, which mainly refers to the <strong>MiRKATMC</strong> function that sets “data.type” to “nominal” in R package, <strong>MiRKATMC</strong>. In addition, we fix a bug in the <strong>MiRKATMC</strong> function, i.e., the p-value combination method HMP reports an error When zero value appears in these combined p-values.</p>
</div>
<div id="example-10" class="section level3">
<h3>5.11.2. Example</h3>
<p><strong>Import requisite R packages</strong></p>
<pre><code>require(ecodist)
require(GUniFrac)
require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Import example microbiome data</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])</code></pre>
<p><strong>Fit MiRKAT-MCN</strong></p>
<pre><code>Ks &lt;- Whole.Kernels(otu.tab, tree, distance = c(&quot;BC&quot;, &quot;U.UniFrac&quot;, &quot;G.UniFrac.00&quot;, &quot;G.UniFrac.25&quot;, &quot;G.UniFrac.50&quot;, &quot;G.UniFrac.75&quot;, &quot;W.UniFrac&quot;), kernel = c(&quot;LK&quot;, &quot;GK&quot;, &quot;LaK&quot;))
out.MiRKAT_MCN &lt;- MiRKAT_MCN(y = y, Ks = Ks, covs = covs)</code></pre>
</div>
</div>
<div id="function-mirkat_mco" class="section level2">
<h2>5.12. Function <strong>MiRKAT_MCO</strong></h2>
<div id="description-11" class="section level3">
<h3>5.12.1. Description</h3>
<p>The function, <strong>MiRKAT_MCO</strong>, is MiRKAT_MC based on cumulative link model, which mainly refers to the <strong>MiRKATMC</strong> function that sets “data.type” to “ordinal” in R package, <strong>MiRKATMC</strong>. In addition, we fix a bug in the <strong>MiRKATMC</strong> function, i.e., the p-value combination method HMP reports an error When zero value appears in these combined p-values.</p>
</div>
<div id="example-11" class="section level3">
<h3>5.12.2. Example</h3>
<p><strong>Import requisite R packages</strong></p>
<pre><code>require(ecodist)
require(GUniFrac)
require(phyloseq)
require(VGAM)
require(MASS)
require(permute)
require(ape)
require(Matrix)
require(multiMiAT)</code></pre>
<p><strong>Import example microbiome data</strong></p>
<pre><code>data(biom1)
otu.tab &lt;- otu_table(biom1)
tree &lt;- phy_tree(biom1)
sample.data &lt;- sample_data(biom1)
y &lt;- sample.data$DiseaseState
y[which(y == &quot;Health&quot;)] = 1
y[which(y == &quot;Adenoma&quot;)] = 2
y[which(y == &quot;Cancer&quot;)] = 3
y &lt;- as.factor(y)
x1 &lt;- sample.data$Age
x2 &lt;- sample.data$Race
x2[which(x2 == &quot;White&quot;)] &lt;- 1
x2[which(x2 == &quot;Other&quot;)] &lt;- 0
x2 &lt;- as.factor(x2)
covs &lt;- as.data.frame(cbind(x1, x2))
colnames(covs) &lt;- c(&quot;Age&quot;, &quot;Race&quot;)
covs[,1] &lt;- as.numeric(covs[,1])
covs[,2] &lt;- as.factor(covs[,2])</code></pre>
<p><strong>Fit MiRKAT-MCO</strong></p>
<pre><code>Ks &lt;- Whole.Kernels(otu.tab, tree, distance = c(&quot;BC&quot;, &quot;U.UniFrac&quot;, &quot;G.UniFrac.00&quot;, &quot;G.UniFrac.25&quot;, &quot;G.UniFrac.50&quot;, &quot;G.UniFrac.75&quot;, &quot;W.UniFrac&quot;), kernel = c(&quot;LK&quot;, &quot;GK&quot;, &quot;LaK&quot;))
out.MiRKAT_MCO &lt;- MiRKAT_MCO(y = y, Ks = Ks, covs = covs)</code></pre>
</div>
</div>
</div>
<div id="dataset" class="section level1">
<h1>6. Dataset</h1>
<p>biom1: Real data on colorectal cancer.</p>
<p>biom2: Real data on <em>Clostridium difficile</em> infections.</p>
<p>phy.parameters: Estimated parameters of Dirichlet-multinomial distribution.</p>
</div>
<div id="references" class="section level1">
<h1>7. References</h1>
<ul>
<li><p>Sun H, et al. multiMiAT: An optimal microbiome-based association test for multicategory phenotypes. (under review)</p></li>
<li><p>Agarwal D and Zhang NR. Semblance: Anempirical similarity kernel on probability spaces. <em>Science Advances</em> 2019;<strong>5</strong>(12):eaau9630.</p></li>
<li><p>Agresti, A. Categorical data analysis (3rd ed). <em>Wiley</em> 2013.</p></li>
<li><p>Bi W, et al. Efficient mixed model approach for large-scale genome-wide association studies of ordinal categorical phenotypes. <em>American Journal of Human Genetics</em> 2021;<strong>108</strong>(5):825–839.</p></li>
<li><p>Bray JR and Curtis JT. An ordination of the upland forest communities of southern Wisconsin. <em>Ecological Monographs</em> 1957;<strong>27</strong>(4):325-349.</p></li>
<li><p>Chen J, et al. Associating microbiome composition with environmental covariates using generalized UniFrac distances. <em>Bioinformatics</em> 2012;<strong>28</strong>(16):2106-2113.</p></li>
<li><p>Chen J and Li H. Variable selection for sparse Dirichlet-multinomial regression with an application to microbiome data analysis. <em>Annals of Applied Statistics</em> 2013;<strong>7</strong>(1).</p></li>
<li><p>Jaccard P. The distribution of the flora in the alpine zone. <em>New Phytologist</em> 1912;<strong>11</strong>(2):37-50.</p></li>
<li><p>Jiang Z, et al. MiRKAT-MC: A distance-based microbiome kernel association test with multi-categorical outcomes. <em>Frontiers in Genetics</em> 2022;<strong>13</strong>:841764.</p></li>
<li><p>Koh H, et al. A powerful microbiome-based association test and a microbial taxa discovery framework for comprehensive association mapping. <em>Microbiome</em> 2017;<strong>5</strong>(1):45.</p></li>
<li><p>Koh H, et al. A distance-based kernel association test based on the generalized linear mixed model for correlated microbiome studies. <em>Frontiers in Genetics</em> 2019;<strong>10</strong>:458.</p></li>
<li><p>Liu M, et al. A method for subtype analysis with somatic mutations. <em>Bioinformatics</em> 2021;<strong>37</strong>(1):50-56.</p></li>
<li><p>Lozupone C, et al. Quantitative and qualitative <span class="math inline">\(\beta\)</span> diversity measures lead to different insights into factors that structure microbial communities. <em>Applied and Environmental Microbiology</em> 2007;<strong>73</strong>(5):1576-1585.</p></li>
<li><p>Lozupone CA and Knight R. UniFrac: a new phylogenetic method for comparing microbial communities. <em>Applied and Environmental Microbiology</em> 2005;<strong>73</strong>(5):1576-1585.</p></li>
<li><p>McMurdie PJ and Holmes S. phyloseq: An R package for reproducible interactive analysis and graphics of microbiome census data. <em>PLoS ONE</em>. 2013;<strong>8</strong>(4):e61217.</p></li>
<li><p>Paradis E, et al. APE: Analyses of phylogenetics and evolution in R language. <em>Bioinformatics</em> 2004;<strong>20</strong>(2):289-290.</p></li>
<li><p>Schubert AM, et al. Microbiome data distinguish patients with <em>Clostridium difficile</em> infection and non-<em>C. difficile</em>-associated diarrhea from healthy controls. <em>mBio</em> 2014;<strong>5</strong>(3):e01021–14.</p></li>
<li><p>Sun H, et al. Detecting sparse microbial association signals adaptively from longitudinal microbiome data based on generalized estimating equations. <em>Briefings in Bioinformatics</em> 2022;<strong>23</strong>(5):bbac149.</p></li>
<li><p>Touloumis A. R package multgee: A generalized estimating equations solver for multinomial responses. <em>Journal of Statistical Software</em> 2015;<strong>64</strong>(8):1-14.</p></li>
<li><p>Touloumis A. Simulating correlated binary and multinomial responses under marginal model specification: The SimCorMultRes Package. <em>The R Journal</em> 2016,<strong>8</strong>(2):79.</p></li>
<li><p>Touloumis A, et al. GEE for multinomial responses using a local odds ratios parameterization. <em>Biom</em> 2013,<strong>69</strong>(3):633–640.</p></li>
<li><p>Wilson DJ. The harmonic mean <em>p</em>-value for combining dependent tests. <em>Proceedings of the National Academy of Sciences</em> 2019;<strong>116</strong>(4):1195–1200.</p></li>
<li><p>Wu C, et al. An adaptive association test for microbiome data. <em>Genome Medicine </em> 2016;<strong>8</strong>(1):56.</p></li>
<li><p>Yee TW, et al. The VGAM package for categorical data analysis. <em>Journal of Statistical Software</em> 2010;<strong>32</strong>(10).</p></li>
<li><p>Zackular JP, et al. The human gut microbiome as a screening tool for colorectal cancer. <em>Cancer Prevention Research</em> 2014;<strong>7</strong>(11):1112–1121.</p></li>
<li><p>Zhan X, et al. A fast small‐sample kernel independence test for microbiome community‐level association analysis. <em>Biometrics</em> 2017;<strong>73</strong>(4):1453-1463.</p></li>
<li><p>Zhao N, et al. Testing in microbiome-profiling studies with MiRKAT, the microbiome regression-based kernel association test. <em>American Journal of Human Genetics</em> 2015;<strong>96</strong>(5):797-807.</p></li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
